pkg/gadgets/interface.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022-2023 The Inspektor Gadget authors
```

Comment:
I totally agree that we should clean this up more and reduce the number of interfaces/functions as much as possible. I'd need to check again whether after all the changes we already made on this PR, it is already possible to reduce/unify more. Most of the rather hacky looking "solutions" were done initially to ensure backwards compatibility and matching test results for the integration tests.

> ```
> - Start(eventHandler any) -> used for trace, profile, top, advise.
> - Stop() (result any, err error) -> as above
> - Run() (result any, err error) -> used only for snapshot 
> ```

It would be great if we could reduce it to that, but afaik we have gadgets that don't use Start/Stop but rather Attach/Detach from where they emit the events, and profile gadgets might for example not return events at all but a summary as blob. If we reduce the interfaces too much, we might lose flexibility and/or need to extend too often again. I think a baseline rule should be: one call for one job, so I wouldn't combine `Start()` with passing an event handler for example.

I think that after using this new workflow for both local-gadget and kubectl-gadget (and we're not bound to backward compatibility with that anymore), we can and should clean up and simplify a lot.

Reply:
Fine, let's keep it as it's and them we can reduce it.